text/0000-attribute-property-rationalization.md

```diff
@@ -0,0 +1,174 @@
+
+```hbs
+<div class="{{color}}">
+  <p class="preview">{{firstName}}</p>
+  <input {{prop value=firstName}} {{on 'keydown' (action "changedName")}} />
+  <input value="Submit" type="Button" {{on 'click' (action "submit")}} />
+</div>
+```
+
+By moving to these new semantics we will shore up long standing confusion in this area, unify HTML serialization and DOM construction, and open the door for first class support for custom elements.
+
+## Motivation
+This proposal:
+
+- brings clarity and [addresses confusion](#appendix-a) as how to set a property on an element
+- enables the usage of [custom elements](https://custom-elements-everywhere.com/)
+- fixes HTML serialization issues for server-side rendering
+
+## Detailed design
+
+To move to this more explicit syntax in a backwards compatible we will convert well-known properties to use the `{{prop}}` and `{{on}}` modifiers via a codemod and retain the runtime setting rules for `attributeBindings`.
+
+### Transforming Through Codemod
+Due to the declarative nature of HTML we are able to see exactly which attributes are being set on which elements. This allows us to understand where we need to set properties and where we can safely rely on setting attributes. Using the list of [HTML attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes) and [WAI-ARIA attributes](https://www.w3.org/TR/wai-aria-1.1/) we can then determine which attributes in templates need to be converted to use either `{{prop}}` or `{{on}}`. We will create a "best effort" codemod to transform all of these cases by default.
+
+### Runtime Rules
+With Ember Components we have the notion of `attributeBindings` to set attributes on the wrapping element as defined by the `tagName` on the component class. For this case we will continue to use the setting rules we have today. Notably, [Glimmer Components](https://glimmerjs.com/guides/templates-and-helpers) have adopted the "Outer HTML" semantics long ago and the experience has been very positive. This means that when Glimmer Components become available in Ember they would be covered by the codemod as described above.
+
+### `{{prop}}` Static Semantics
+
+- {{__prop__ _HashPair_}}
+  - {{__prop__ _HashPair_ ... }}
+
+- _HashPair_ __:__
+  - _Key_ __=__ _Value_
+
+- _Key_ __:__
+  - _PathExpression_
+  - _StringLiteral_
+
+- _Value_ __:__
+  - _PathExpression_
+  - _StringLiteral_
+  - _BooleanLiteral_
+  - _NullLiteral_
+  - _UndefinedLiteral_
+  - _NumberLiteral_
+
+### `{{prop}}` Runtime Semantics
+The `{{prop}}` modifier is installed on the element once the element is placed in the DOM. At that time it will loop over the `HashPair`(s) setting the properties on to the element. No normalization is done for the `Key` or `Value`. In other words what you see is what you get.
```

Comment:
The timing is problematic for custom elements.

Custom elements (web components) have lifecycle hooks similar to `@ember/component`. One of them is `connectedCallback`. It's executed when the element is insert into the DOM.

The property is not available in that hook if `{{prop}}` modifier runs after the element had been insert into DOM.

I'm aware that this is a limitation of modifiers in general and not related to `{{prop}}` modifier proposed here only. The limitation is discussed in detail at #652.